core/src/main/java/org/neo4j/graphalgo/core/utils/LongMinPriorityQueue.java

```diff
@@ -0,0 +1,198 @@
+        if (size > 0) {
+            long result = heap[1];       // save first value
+            heap[1] = heap[size];     // move last to first
+            size--;
+            downHeap(1);              // adjust heap
+            costs.remove(result);
+            return result;
+        } else {
+            return -1;
+        }
+    }
+
+    /**
+     * @return the number of elements currently stored in the queue.
+     */
+    public int size() {
+        return size;
+    }
+
+    /**
+     * @return true iff there are currently no elements stored in the queue.
+     */
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    /**
+     * @return true iff there is currently at least one element stored in the queue.
+     */
+    public boolean nonEmpty() {
+        return size != 0;
+    }
+
+    /**
+     * Removes all entries from the queue.
+     */
+    public void clear() {
+        size = 0;
+    }
+
+    /**
+     * Removes all entries from the queue, releases all buffers.
+     * The queue can no longer be used afterwards.
+     */
+    public void release() {
+        size = 0;
+        heap = null;
+
+        costs.keys = new long[0];
+        costs.clear();
+        costs.keys = null;
+        costs.values = null;
+    }
+
+    private boolean upHeap(int origPos) {
+        int i = origPos;
+        long node = heap[i];          // save bottom node
+        int j = i >>> 1;
+        while (j > 0 && lessThan(node, heap[j])) {
+            heap[i] = heap[j];       // shift parents down
+            i = j;
+            j = j >>> 1;
+        }
+        heap[i] = node;            // install saved node
+        return i != origPos;
+    }
+
+    private void downHeap(int i) {
+        long node = heap[i];          // save top node
+        int j = i << 1;              // find smaller child
+        int k = j + 1;
+        if (k <= size && lessThan(heap[k], heap[j])) {
+            j = k;
+        }
+        while (j <= size && lessThan(heap[j], node)) {
+            heap[i] = heap[j];       // shift up child
+            i = j;
+            j = i << 1;
+            k = j + 1;
+            if (k <= size && lessThan(heap[k], heap[j])) {
+                j = k;
+            }
+        }
+        heap[i] = node;            // install saved node
+    }
+
+    private void ensureCapacityForInsert() {
+        if (size >= heap.length) {
+            long[] newHeap = new long[ArrayUtil.oversize(
+                    size + 1,
+                    Integer.BYTES)];
+            System.arraycopy(heap, 0, newHeap, 0, heap.length);
+            heap = newHeap;
+        }
+    }
+
+    @Override
+    public PrimitiveLongIterator iterator() {
+        return new PrimitiveLongIterator() {
+
+            int i = 1;
+
+            @Override
+            public boolean hasNext() {
+                return i <= size;
+            }
+
+            @Override
+            public long next() {
+                if (!hasNext()) {
```

Comment:
don't need to check every time, the array itself has an overflow check which will throw the exception